### PR TITLE
Entity GraphQL type rework

### PR DESCRIPTION
**Pull request by thehabbos007**
## 🌟 What is the purpose of this PR?

This PR rewrites the `PersistedEntity` and its implementors to match the new shape from the Graph.
Some of the names haven't been renamed yet (Persisted is still part of the Knowledge GQL types).
I've also deleted old GQL types we are not directly using anymore (`Page`).

The new type for Entity is called `EntityWithMetadata` and is still an interface. The idea is that we are using the same shape that we would return from the backend, while allowing custom resolvers for backwards compatibility while we rewrite frontend bits to use subgraphs.

When we get around to deprecating/removing the legacy GQL types that share the "Entity" name, we will be able to remove the `WithMetadata` suffix and have the name be consistent with the rest of the system. The other knowledge system types have not been renamed yet, though.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203150072758509/1203329839349733/f) _(internal)_
- [Asana task](https://app.asana.com/0/1203150072758509/1203329839349728/f) _(internal)_

## 🔍 What does this change?

- See commits

## 📜 Does this require a change to the docs?

- May need to rework the GQL documentation strings

## ⚠️ Known issues

- The queries consuming the types have not been adjusted, but codegen runs (maybe? turborepo seems to cache the state for me locally, so not sure if there are problems?)
- Resolvers are broken now

### CODE REVIEW

File: packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts
Comment (thehabbos007) [inline packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts:20]: We will only need this because of the way we use the `EntityWithMetadata` interface. We can hopefully get rid of this as we move to using the subgraph for the frontend.

File: packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts
Comment (thehabbos007) [inline packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts:55]: We probably don't want to get rid of this yet as we need it for page saving

File: packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts
Comment (thehabbos007) [inline packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts:26]: I moved these legacy things out of page to persist the pagination typedefs. We don't really need this, and this entire file and its resolver could be deleted as well.

File: packages/hash/api/src/graphql/resolvers/index.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/graphql/resolvers/index.ts:83]: Are we naming these to avoid conflicts for now? Because ideally these would just be `createEntity` and such, right?

File: packages/hash/api/src/graphql/resolvers/index.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/graphql/resolvers/index.ts:258]:
```suggestion
     * Determines whether an `EntityWithMetadata` instance should be treated as a
```

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/index.ts:258]: Do we still need this thing anyway?

File: packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts:32]: What does this even do? Do we still need it, the docs and such don't make sense to me

File: packages/hash/api/src/graphql/typeDefs/knowledge/page.typedef.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/graphql/typeDefs/knowledge/page.typedef.ts:4]: Do we still need to implement?

File: packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts:26]: Why keep it then?

File: packages/hash/frontend/src/shared/layout/layout-with-header/search-bar.tsx
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/shared/layout/layout-with-header/search-bar.tsx:1]: Should we move this down to the commented out code (can always move it back up when uncommenting) or just delete it?

File: packages/hash/frontend/src/shared/layout/layout-with-header/search-bar.tsx
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/shared/layout/layout-with-header/search-bar.tsx:161]: Is there an asana task we can associate with this? I remember seeing some for search

File: packages/hash/subgraph/src/types/element.ts
Comment (Alfred-Mountfield) [inline packages/hash/subgraph/src/types/element.ts:49]: Thanks for doing this, @benwerner01 had a use for it in his PR too. We could of course always just do `Entity["metadata"]` but I think this is probably nicer 👍

**Review by Alfred-Mountfield — COMMENTED** [review]
Generally looks good to me! I think @benwerner01 is going to have to review this too. (I do have some comments in the mean time)

I think I don't have a good sense of how much is changing with us deleting all of this, but we can maybe figure that out in follow-ups, and bring stuff back if we really need to,

File: packages/hash/api/src/graphql/resolvers/index.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/index.ts:83]: Yes, some of them we could get away with calling just `Entity`, but for consistency after deprecating/removing the old ones, we can simply search/replace `WithMetadata`

File: packages/hash/api/src/graphql/resolvers/index.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/resolvers/index.ts:258]: Unfortunately yes, until we move to subgraphs exclusively on the frontend, this needs to be there for the different implementers of the Entity interface

File: packages/hash/api/src/graphql/resolvers/index.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/resolvers/index.ts:83]: I think that makes sense to not get stuck in refactoring hell. cc @CiaranMn

File: packages/hash/api/src/graphql/typeDefs/knowledge/page.typedef.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/knowledge/page.typedef.ts:4]: For now, yes. But hopefully we can get rid of this when frontend relies on subgraphs. Only reason why this is kept is so that pages and comments would not need a full-on frontend rewrite yet.

File: packages/hash/api/src/graphql/resolvers/index.ts
Comment (CiaranMn) [reply packages/hash/api/src/graphql/resolvers/index.ts:83]: Let's make getting rid of old stuff something we do immediately after merging to main

File: packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts:26]: I suppose we could get rid of it, will see how much it breaks with codegen

File: packages/hash/api/src/graphql/typeDefs/knowledge/page.typedef.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/knowledge/page.typedef.ts:4]: I more meant that can we still keep their structure (with the interface fields 🙃) but not actually `implements`. If we're using the interface fair enough.

File: packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts:26]: Please do, thanks!

File: packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts
Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts:32]: Would appreciate you looking into that, thanks

File: packages/hash/api/src/graphql/typeDefs/knowledge/page.typedef.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/knowledge/page.typedef.ts:4]: I'll give it another look and check these relationships. Maybe we can get rid of `UnknownEntity`

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/knowledge/page.typedef.ts:4]: Would be fantastic

File: packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/paginationConnections.typedef.ts]: Removed in a5f80242a582547c9a976c1b885e37721193731c

File: packages/hash/frontend/src/shared/layout/layout-with-header/search-bar.tsx
Comment (thehabbos007) [reply packages/hash/frontend/src/shared/layout/layout-with-header/search-bar.tsx:164]: Added reference to the task in a5f80242a582547c9a976c1b885e37721193731c

File: packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts]: I think I am able to get rid of this `entityTypeChoice` entirely, see 655d3221a266fdfd72cbc0e6cf6d2fd498022a42.
Page saving is only using the `entityTypeId` which is the only option anyway

File: packages/hash/api/src/graphql/typeDefs/knowledge/page.typedef.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/knowledge/page.typedef.ts]: I got rid of unknown entirely in 61cd88133749f9c94709511cc2460d2925a27aca
This only works because we don't rely on the polymorphism of types anymore since we introduced Subgraph for all the parts of the queries that would require being able to tell what type we're dealing with

File: packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts
Comment (Alfred-Mountfield) [inline packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts:38]: Shouldn't this be `EntityType`?

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts:38]: Can you update as necessary then :)

Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts:38]: String would be correct here. We don't have a `VerisonedUri` type/scalar. yet Do we want to introduce it here? I can export it from the type system package from the subgraph lib to get the template literal string

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts:38]: I think all we want to do is rename the variable if that's the case. (Just `EntityTypeId` or something)

We should probably consider adding `VersionedUri` to the GraphQL types soon though, as otherwise we are going to have to mess around with casts or checks on either side of the boundary which will be a pain. I leave it up to you to decide if it's big enough to constitute its own task :)

Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts]: Introduced `VersionedUri` 9607d6d047f77b25db9016fa492c52d902bf98f7 and adjusted references

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts]: I think codegen is broken on that commit?

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts]: > @hashintel/hash-frontend:codegen: [FAILED]   Error 0: Variable "$dataTypeId" of type "String!" used in position expecting type "VersionedUri!".

Comment (Alfred-Mountfield) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts]: Fixed it locally on my branch while experimenting so I've pushed :) We'll hopefully get this turborepo caching problem fixed

File: packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts
Comment (benwerner01) [inline packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts:9]: Do we intend on just using the `Entity` scalar down the line? If so can we leave a `@todo` here? If not, why not?

File: packages/hash/frontend/src/graphql/queries/knowledge/entity.queries.ts
Comment (benwerner01) [inline packages/hash/frontend/src/graphql/queries/knowledge/entity.queries.ts:105]: This mutation doesn't exist anymore, did you intend on replacing it with `updateEntityWithMetadata`?

**Review by benwerner01 — COMMENTED** [review]
Partial Review

File: packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts:9]: I think we should, yeah. I'll leave a todo. The reason why we aren't doing it yet is to allow for custom resolvers for pages/comments/etc. to allow migrating in pieces before switching to subgraphs

Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts:9]: Now that I think about it, we don't need this type to be here as it's not an interface type anymore

File: packages/hash/shared/codegen.yml
Comment (thehabbos007) [reply packages/hash/shared/codegen.yml:34]: Nope, collab isn't used to any extent in the current main branch. I will remove the line instead of commenting this out to make that explicit. Collab is even removed from TS compilation in the API tsconfig.

File: packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts
Comment (thehabbos007) [reply packages/hash/api/src/graphql/typeDefs/knowledge/entity.typedef.ts]: I've addressed the todo and removed the explicit type in 029b8ed033c5361bc4e7b630bdb33b6db83c995a

File: packages/hash/frontend/src/graphql/queries/knowledge/entity.queries.ts
Comment (thehabbos007) [reply packages/hash/frontend/src/graphql/queries/knowledge/entity.queries.ts:76]: Yes, thanks this is intended to be replaced. I'm very confused as to why this isn't being picked up by codegen?
Anyway, I will update these but I expect the follow-up tasks to fix the frontend would make sure these queries are updated correctly: 3ec652e69ae70c16980a7b8a4e1b7ae13f803299

Comment (benwerner01) [reply packages/hash/frontend/src/graphql/queries/knowledge/entity.queries.ts:76]: > I'm very confused as to why this isn't being picked up by codegen?

To be honest, me too - is this potentially caching related as well?

File: packages/hash/shared/codegen.yml
Comment (benwerner01) [reply packages/hash/shared/codegen.yml]: We should then move all GQL queries out of the `shared` package, right? If so perhaps this is too big for this PR, but we should probably leave a `@todo` and create a task.

File: packages/hash/frontend/src/graphql/queries/knowledge/entity.queries.ts
Comment (thehabbos007) [reply packages/hash/frontend/src/graphql/queries/knowledge/entity.queries.ts:76]: I don't think it's caching, running `yarn  graphql-codegen --config codegen.yml` from the frontend folder seems to provide the same output. Perhaps the query isn't used anywhere?

File: packages/hash/shared/codegen.yml
Comment (thehabbos007) [reply packages/hash/shared/codegen.yml]: The queries related to page were moved from `collab` to `shared` when we switched to local page-saving, are you suggesting we move them from `shared` to `frontend`? They are only consumed by frontend currently, so that would be correct, but not sure if it should be done here or in one of the PRs coming up for fixing the frontend

**Review by benwerner01 — APPROVED** [review]

**Pull request merged**
